README.md
# Lockit forgot password

[![Build Status](https://travis-ci.org/zemirco/lockit-forgot-password.svg?branch=master)](https://travis-ci.org/zemirco/lockit-forgot-password)
[![NPM version](https://badge.fury.io/js/lockit-forgot-password.svg)](http://badge.fury.io/js/lockit-forgot-password)
[![Dependency Status](https://david-dm.org/zemirco/lockit-forgot-password.svg)](https://david-dm.org/zemirco/lockit-forgot-password)

Help users who forgot their passwords for your Express app. The module is part of [Lockit](https://github.com/zemirco/lockit).

## Installation

`npm install lockit-forgot-password`

```js
var ForgotPassword = require('lockit-forgot-password');
var utils = require('lockit-utils');
var config = require('./config.js');

var db = utils.getDatabase(config);
var adapter = require(db.adapter)(config);

var app = express();

// express settings
// ...
// sessions are required - either cookie or some sort of db
app.use(cookieParser());
app.use(cookieSession({
  secret: 'this is my super secret string'
}));

// create new ForgotPassword instance
var forgotPassword = new ForgotPassword(config, adapter);

// use forgotPassword.router with your app
app.use(forgotPassword.router);
```

## Configuration

More about configuration at [Lockit](https://github.com/zemirco/lockit).

## Features

 - allow password reset for users
 - input validation
 - link expiration times
 - user email verification via unique token
 - hash password using [pbkdf2](http://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2_password_salt_iterations_keylen_callback)
 - token format verification before database querying

## Routes included

 - GET /forgot-password
 - POST /forgot-password
 - GET /forgot-password/:token
 - POST /forgot-password/:token

## REST API

If you've set `exports.rest` in your `config.js` the module behaves as follows.

 - all routes have `/rest` prepended
 - `GET /rest/forgot-password` is `next()`ed and you can catch `/forgot-password` on the client
 - `POST /rest/forgot-password` stays the same but sends JSON
 - `GET /rest/forgot-password/:token` sends JSON and you can catch `/forgot-password/:token` on the client
 - `POST /rest/forgot-password/:token` sends JSON

## Test

`$ npm test`

## License

MIT


History.md

##### 1.3.2 / 2015-07-06

- update dependencies

##### 1.3.1 / 2015-06-30

- update dependencies
- use make instead grunt
- use eslint
- add node 0.12 to travis

##### 1.3.0 / 2014-09-27

- update dependencies

##### 1.2.0 / 2014-07-23

- add events `'forgot::sent'` and `'forgot::success'`
- update dependencies

##### 1.1.1 / 2014-05-27

- set `autocomplete="off"`
- use Bootstrap responsive classes

##### 1.1.0 / 2014-05-23

- refactor code
- update dependencies
- use updated [lockit-sendmail](https://github.com/zeMirco/lockit-sendmail)

##### 1.0.0 / 2014-04-19

- requires Express 4.x
- makes use of `express.Router()`. No need to pass `app` around as argument.

  **old**

  ```js
  var ForgotPassword = require('lockit-forgot-password');

  var forgotPassword = new ForgotPassword(app, config, adapter);
  ```

  **new**

  ```js
  var ForgotPassword = require('lockit-forgot-password');

  var forgotPassword = new ForgotPassword(config, adapter);
  app.use(forgotPassword.router);
  ```

- proper Error handling. All Errors are piped to next middleware.

  **old**

  ```js
  if (err) console.log(err);
  ```

  **new**

  ```js
  if (err) return next(err);
  ```

  Make sure you have some sort of error handling middleware at the end of your
  routes (is included by default in Express 4.x apps if you use the `express-generator`).

##### 0.5.0 / 2014-04-11

- `username` becomes `name`


test/app.js

/**
 * Module dependencies.
 */

var express = require('express');
var routes = require('./routes');
var user = require('./routes/user');
var http = require('http');
var path = require('path');
var fs = require('fs');

// require delete account middleware
//var config = require('./config.js');
var forgotPassword = require('../index.js');

function start(config) {

  config = config || require('./config.js');

  var app = express();

// set basedir so views can properly extend layout.jade
  app.locals.basedir = __dirname + '/views'; // comment out and error returns

// all environments
  app.set('port', process.env.PORT || config.port || 3000);
  app.set('views', __dirname + '/views');
  app.set('view engine', 'jade');
  // make JSON output simpler for testing
  app.set('json spaces', 0);
  app.use(express.favicon());
  app.use(express.bodyParser());
  app.use(express.methodOverride());
  app.use(express.cookieParser('your secret here'));
  app.use(express.cookieSession());

// set a dummy session for testing purpose
  app.use(function(req, res, next) {
    req.session.username = 'john';
    next();
  });

// use forgot password middleware with testing options
  forgotPassword(app, config);

  app.use(app.router);
  app.use(express.static(path.join(__dirname, 'public')));

// development only
  if ('development' == app.get('env')) {
    app.use(express.errorHandler());
  }

  app.get('/', routes.index);
  app.get('/users', user.list);

  http.createServer(app).listen(app.get('port'), function(){
    console.log('Express server listening on port ' + app.get('port'));
  });

  return app;

}

// export app for testing
if(require.main === module){
  // called directly
  start();
} else {
  // required as a module -> from test file
  module.exports = function(config) {
    return start(config);
  };
}

test/csrf.js
'use strict';

var request = require('supertest');
var should = require('should'); // eslint-disable-line no-unused-vars
var cookie = require('cookie');
var utls = require('lockit-utils');

var config = require('./app/config.js');
var app = require('./app/app.js');

var db = utls.getDatabase(config);
var adapter = require(db.adapter)(config);

var _config = JSON.parse(JSON.stringify(config));
_config.port = 9000;
_config.csrf = true;
_config.forgotPassword.tokenExpiration = '10 ms';
var _app = app(_config);

// second app without csrf so we can easily make a POST request to create token
var configWithoutCSRF = JSON.parse(JSON.stringify(config));
configWithoutCSRF.port = 9001;
configWithoutCSRF.forgotPassword.tokenExpiration = '1 hour';
var appWithoutCSRF = app(configWithoutCSRF);

describe('# csrf', function() {

  before(function(done) {
    adapter.save('csrf', 'csrf@email.com', 'password', function() {
      adapter.find('name', 'csrf', function(err, user) {
        if (err) {console.log(err); }
        user.emailVerified = true;
        adapter.update(user, done);
      });
    });
  });

  describe('GET /forgot-password', function() {

    it('should include the token in the view', function(done) {
      request(_app)
        .get('/forgot-password')
        .end(function(err, res) {
          if (err) {console.log(err); }
          var cookies = cookie.parse(res.headers['set-cookie'][0]);
          var token = cookies.csrf;
          res.text.should.containEql('name="_csrf" value="' + token + '"');
          done();
        });
    });

  });

  describe('GET /forgot-password/:token', function() {

    it('should include the token in the view', function(done) {
      // create token
      request(appWithoutCSRF)
        .post('/forgot-password')
        .send({email: 'csrf@email.com'})
        .end(function() {
          // find token
          adapter.find('name', 'csrf', function(err, user) {
            if (err) {console.log(err); }
            request(_app)
              .get('/forgot-password/' + user.pwdResetToken)
              .end(function(error, res) {
                if (error) {console.log(error); }
                var cookies = cookie.parse(res.headers['set-cookie'][0]);
                var token = cookies.csrf;
                res.text.should.containEql('name="_csrf" value="' + token + '"');
                done();
              });
          });
        });
    });

  });

  after(function(done) {
    adapter.remove('csrf', done);
  });

});


test/rest.js
'use strict';

var request = require('supertest');
var should = require('should'); // eslint-disable-line no-unused-vars
var uuid = require('node-uuid');
var utls = require('lockit-utils');

var config = require('./app/config.js');
var app = require('./app/app.js');

var db = utls.getDatabase(config);
var adapter = require(db.adapter)(config);

// testing the rest api
var confiRest = JSON.parse(JSON.stringify(config));
confiRest.port = 6000;
confiRest.rest = true;
var appRest = app(confiRest);

var configTokenExpiration = JSON.parse(JSON.stringify(config));
configTokenExpiration.port = 6001;
configTokenExpiration.forgotPassword.tokenExpiration = '1 ms';
var appTokenExpiration = app(configTokenExpiration);

describe('# rest enabled', function() {

  // add a dummy user to db
  before(function(done) {
    adapter.save('rest', 'rest@email.com', 'password', function() {
      adapter.save('alan', 'alan@email.com', 'password', done);
    });
  });

  describe('GET /forgot-password', function() {

    it('should be forwarded to error handling middleware when rest is active', function(done) {
      request(appRest)
        .get('/rest/forgot-password')
        .end(function(err, res) {
          if (err) {console.log(err); }
          res.statusCode.should.equal(404);
          done();
        });
    });

  });

  describe('POST /forgot-password', function() {

    it('should return an error when email has invalid format', function(done) {
      request(appRest)
        .post('/rest/forgot-password')
        .send({email: 'someemail.com'})
        .end(function(error, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Email is invalid"}');
          done();
        });
    });

    it('should render a success message when no user was found', function(done) {
      request(appRest)
        .post('/rest/forgot-password')
        .send({email: 'jim@wayne.com'})
        .end(function(error, res) {
          res.statusCode.should.equal(204);
          done();
        });
    });

    it('should render a success message when email was sent', function(done) {
      request(appRest)
        .post('/rest/forgot-password')
        .send({email: 'rest@email.com'})
        .end(function(error, res) {
          res.statusCode.should.equal(204);
          done();
        });
    });

  });

  describe('GET /forgot-password/:token', function() {

    it('should forward to error handling middleware when token has invalid format', function(done) {
      request(appRest)
        .get('/rest/forgot-password/some-test-token-123')
        .end(function(err, res) {
          if (err) {console.log(err); }
          res.statusCode.should.equal(404);
          done();
        });
    });

    it('should forward to error handling middleware when no user for token is found', function(done) {
      var token = uuid.v4();
      request(appRest)
        .get('/rest/forgot-password/' + token)
        .end(function(err, res) {
          if (err) {console.log(err); }
          res.statusCode.should.equal(404);
          done();
        });
    });

    it('should render the link expired template when token has expired', function(done) {
      // create token
      request(appTokenExpiration)
        .post('/forgot-password')
        .send({email: 'alan@email.com'})
        .end(function() {
          // get token from db
          adapter.find('name', 'alan', function(error, user) {
            if (error) {console.log(error); }
            // use GET request
            request(appRest)
              .get('/rest/forgot-password/' + user.pwdResetToken)
              .end(function(err, res) {
                if (err) {console.log(err); }
                res.statusCode.should.equal(403);
                res.text.should.equal('{"error":"link expired"}');
                done();
              });
          });
        });
    });

    it('should render a form to enter the new password', function(done) {
      // create token
      request(appRest)
        .post('/forgot-password')
        .send({email: 'rest@email.com'})
        .end(function() {
          // get token from db
          adapter.find('name', 'rest', function(err, user) {
            if (err) {console.log(err); }
            // use GET request
            request(appRest)
              .get('/rest/forgot-password/' + user.pwdResetToken)
              .end(function(error, res) {
                if (error) {console.log(error); }
                res.statusCode.should.equal(204);
                done();
              });
          });
        });
    });

  });

  describe('POST /forgot-password/:token', function() {

    it('should return with an error message when password is empty', function(done) {
      var token = uuid.v4();
      request(appRest)
        .post('/rest/forgot-password/' + token)
        .send({password: ''})
        .end(function(err, res) {
          if (err) {console.log(err); }
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Please enter a password"}');
          done();
        });
    });

    it('should render the link expired template when token has expired', function(done) {
      // create token
      request(appTokenExpiration)
        .post('/forgot-password')
        .send({email: 'alan@email.com'})
        .end(function() {
          // get token from db
          adapter.find('name', 'alan', function(err, user) {
            if (err) {console.log(err); }
            // use token from db for POST request
            request(appRest)
              .post('/rest/forgot-password/' + user.pwdResetToken)
              .send({password: 'something'})
              .end(function(error, res) {
                if (error) {console.log(error); }
                res.statusCode.should.equal(403);
                res.text.should.equal('{"error":"link expired"}');
                done();
              });
          });
        });
    });

    it('should render a success message when everything is fine', function(done) {
      // create token
      request(appRest)
        .post('/forgot-password')
        .send({email: 'rest@email.com'})
        .end(function() {
          // get token from db
          adapter.find('name', 'rest', function(err, user) {
            if (err) {console.log(err); }
            // use token from db for POST request
            request(appRest)
              .post('/rest/forgot-password/' + user.pwdResetToken)
              .send({password: 'something'})
              .end(function(error, res) {
                if (error) {console.log(error); }
                res.statusCode.should.equal(204);
                done();
              });
          });
        });
    });

  });

  after(function(done) {
    adapter.remove('rest', function() {
      adapter.remove('alan', done);
    });
  });

});


index.js
'use strict';

var path = require('path');
var events = require('events');
var util = require('util');
var express = require('express');
var uuid = require('node-uuid');
var pwd = require('couch-pwd');
var ms = require('ms');
var moment = require('moment');
var Mail = require('lockit-sendmail');


/**
 * Internal helper functions
 */
function join(view) {
  return path.join(__dirname, 'views', view);
}



/**
 * ForgotPassword constructor function.
 *
 * @param {Object} config
 * @param {Object} adapter
 */
var ForgotPassword = module.exports = function(config, adapter) {

  if (!(this instanceof ForgotPassword)) {return new ForgotPassword(config, adapter); }

  // call super constructor function
  events.EventEmitter.call(this);

  this.config = config;
  this.adapter = adapter;

  // set default route
  var route = config.forgotPassword.route || '/forgot-password';

  // add prefix when rest is active
  if (config.rest) {route = '/rest' + route; }

  /**
   * Routes
   */

  var router = new express.Router();
  router.get(route, this.getForgot.bind(this));
  router.post(route, this.postForgot.bind(this));
  router.get(route + '/:token', this.getToken.bind(this));
  router.post(route + '/:token', this.postToken.bind(this));
  this.router = router;

};

util.inherits(ForgotPassword, events.EventEmitter);



/**
 * GET /forgot-password
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
ForgotPassword.prototype.getForgot = function(req, res, next) {
  var config = this.config;

  // do not handle the route when REST is active
  if (config.rest) {return next(); }

  // custom or built-in view
  var view = config.forgotPassword.views.forgotPassword || join('get-forgot-password');

  res.render(view, {
    title: 'Forgot password',
    basedir: req.app.get('views')
  });
};



/**
 * POST /forgot-password
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
ForgotPassword.prototype.postForgot = function(req, res, next) {
  var config = this.config;
  var adapter = this.adapter;
  var that = this;

  var email = req.body.email;

  var error = null;
  // regexp from https://github.com/angular/angular.js/blob/master/src/ng/directive/input.js#L4
  var EMAIL_REGEXP = /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,6}$/;

  // check for valid input
  if (!email || !email.match(EMAIL_REGEXP)) {
    error = 'Email is invalid';

    // send only JSON when REST is active
    if (config.rest) {return res.json(403, {error: error}); }

    // custom or built-in view
    var errorView = config.forgotPassword.views.forgotPassword || join('get-forgot-password');

    res.status(403);
    return res.render(errorView, {
      title: 'Forgot password',
      error: error,
      basedir: req.app.get('views')
    });
  }

  // looks like given email address has the correct format

  // look for user in db
  adapter.find('email', email, function(err, user) {
    if (err) {return next(err); }

    // custom or built-in view
    var view = config.forgotPassword.views.sentEmail || join('post-forgot-password');

    // no user found -> pretend we sent an email
    if (!user) {
      // send only JSON when REST is active
      if (config.rest) {return res.send(204); }

      return res.render(view, {
        title: 'Forgot password',
        basedir: req.app.get('views')
      });
    }

    // user found in db
    // do not delete old password as it might be someone else
    // send link with setting new password page
    var token = uuid.v4();
    user.pwdResetToken = token;

    // set expiration date for password reset token
    var timespan = ms(config.forgotPassword.tokenExpiration);
    user.pwdResetTokenExpires = moment().add(timespan, 'ms').toDate();

    // update user in db
    adapter.update(user, function(er, usr) {
      if (er) {return next(er); }

      // send email with forgot password link
      var mail = new Mail(config);
      mail.forgot(usr.name, usr.email, token, function(e) {
        if (e) {return next(e); }

        // emit event
        that.emit('forgot::sent', usr, res);

        // send only JSON when REST is active
        if (config.rest) {return res.send(204); }

        res.render(view, {
          title: 'Forgot password',
          basedir: req.app.get('views')
        });
      });

    });

  });
};



/**
 * GET /forgot-password/:token
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
ForgotPassword.prototype.getToken = function(req, res, next) {
  var config = this.config;
  var adapter = this.adapter;

  // get token from url
  var token = req.params.token;

  // verify format of token
  var re = new RegExp('[0-9a-f]{22}|[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}', 'i');

  // if format is wrong no need to query the database
  if (!re.test(token)) {return next(); }

  // check if we have a user with that token
  adapter.find('pwdResetToken', token, function(err, user) {
    if (err) {return next(err); }

    // if no user is found forward to error handling middleware
    if (!user) {return next(); }

    // check if token has expired
    if (new Date(user.pwdResetTokenExpires) < new Date()) {
      // make old token invalid
      delete user.pwdResetToken;
      delete user.pwdResetTokenExpires;

      // update user in db
      return adapter.update(user, function(error) {
        if (error) {return next(error); }

        // send only JSON when REST is active
        if (config.rest) {return res.json(403, {error: 'link expired'}); }

        // custom or built-in view
        var view = config.forgotPassword.views.linkExpired || join('link-expired');

        // tell user that link has expired
        res.render(view, {
          title: 'Forgot password - Link expired',
          basedir: req.app.get('views')
        });

      });

    }

    // send only JSON when REST is active
    if (config.rest) {return res.send(204); }

    // custom or built-in view
    var view = config.forgotPassword.views.newPassword || join('get-new-password');

    // render success message
    res.render(view, {
      token: token,
      title: 'Choose a new password',
      basedir: req.app.get('views')
    });

  });
};



/**
 * POST /forgot-password/:token
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
ForgotPassword.prototype.postToken = function(req, res, next) {
  var config = this.config;
  var adapter = this.adapter;
  var that = this;

  var password = req.body.password;
  var token = req.params.token;

  var error = '';

  // verify format of token
  var re = new RegExp('[0-9a-f]{22}|[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}', 'i');

  // if format is wrong no need to query the database
  if (!re.test(token)) {return next(); }

  // check for valid input
  if (!password) {
    error = 'Please enter a password';

    // send only JSON when REST is active
    if (config.rest) {return res.json(403, {error: error}); }

    // custom or built-in view
    var view = config.forgotPassword.views.forgotPassword || join('get-forgot-password');

    res.status(403);
    return res.render(view, {
      title: 'Choose a new password',
      error: error,
      token: token,
      basedir: req.app.get('views')
    });
  }

  // check for token in db
  adapter.find('pwdResetToken', token, function(err, user) {
    if (err) {return next(err); }

    // if no token is found forward to error handling middleware
    if (!user) {return next(); }

    // check if token has expired
    if (new Date(user.pwdResetTokenExpires) < new Date()) {
      // make old token invalid
      delete user.pwdResetToken;
      delete user.pwdResetTokenExpires;

      // update user in db
      return adapter.update(user, function(e) {
        if (e) {return next(e); }

        // send only JSON when REST is active
        if (config.rest) {return res.json(403, {error: 'link expired'}); }

        // custom or built-in view
        var viewExpired = config.forgotPassword.views.linkExpired || join('link-expired');

        // tell user that link has expired
        res.render(viewExpired, {
          title: 'Forgot password - Link expired',
          basedir: req.app.get('views')
        });

      });

    }

    // if user comes from couchdb it has an 'iterations' key
    if (user.iterations) {pwd.iterations(user.iterations); }

    // create hash for new password
    pwd.hash(password, function(hashError, salt, hash) {
      if (hashError) {return next(hashError); }

      // update user's credentials
      user.salt = salt;
      user.derived_key = hash; // eslint-disable-line camelcase

      // remove helper properties
      delete user.pwdResetToken;
      delete user.pwdResetTokenExpires;

      // update user in db
      adapter.update(user, function(updateErr, updateUser) {
        if (updateErr) {return next(updateErr); }

        // emit event
        that.emit('forgot::success', updateUser, res);

        // send only JSON when REST is active
        if (config.rest) {return res.send(204); }

        // custom or built-in view
        var viewChanged = config.forgotPassword.views.changedPassword || join('change-password-success');

        // render success message
        res.render(viewChanged, {
          title: 'Password changed',
          basedir: req.app.get('views')
        });

      });

    });
  });
};
